### ISSUE TITLE
fromFilePairs() and splitFastq() should play well together

### ISSUE THREAD

**Issue by jasonsydes** [nextflow-io/nextflow#672]
Hello there!  

I'm new to Nextflow and am excited by its potential usefulness.  Thank you for developing Nextflow!  I'm also still trying to learn how Nextflow works.

I saw Issue #436, and figured I should be able to do this:
```nextflow
params.reads = "/path/to/test_R{1,2}.fq"

Channel
    .fromFilePairs( params.reads )
    .splitFastq (by: 1, pe: true)
    .set { read_pairs_ch }

read_pairs_ch.println { "Received this $it" }
```
However, no matter what variation I attempt, I always end up with this:
```
N E X T F L O W  ~  version 0.29.0
Launching `./gsnap3.nf` [drunk_hodgkin] - revision: 119155c95f
ERROR ~ Cannot find splittable file (elem=-1)

 -- Check '.nextflow.log' file for details
ERROR ~ Cannot find splittable file (elem=-2)

 -- Check '.nextflow.log' file for details
```
The .nextflow.log says basically the same thing:
```
Apr-28 21:53:04.133 [main] DEBUG nextflow.script.ScriptRunner - > Launching execution
Apr-28 21:53:04.147 [main] DEBUG nextflow.Channel - files for syntax: glob; folder: /path/to; pattern: test_R{1,2}.fq; options: null
Apr-28 21:53:04.217 [Actor Thread 2] ERROR nextflow.extension.DataflowHelper - @unknown
java.lang.IllegalArgumentException: Cannot find splittable file (elem=-1)
```
Instead, I ended up doing something like the following:
```nextflow
params.reads = "/path/to/test_R{1,2}.fq"

Channel
    .fromFilePairs(params.reads)
    .set { sample_id_and_read_pairs_ch }

process foo {
  input:
  set val(sample_id), file(reads) from sample_id_and_read_pairs_ch
  output:
  file reads into read_pairs_ch
  """
  """
}

process bar {
  input:
  set file(R1), file(R2) from read_pairs_ch

  output:
  file R1 into R1_ch
  file R2 into R2_ch
  """
  """
}

R1_ch.splitFastq( by: 1 ).set { R1_split_ch }
R2_ch.splitFastq( by: 1 ).set { R2_split_ch }

R1_split_ch
    .merge(R2_split_ch)
    .set { read_pairs_split_ch }

read_pairs_split_ch.println { "\nReceived this: $it" }
```
Doing that, it looks good at first, because I received the following output:
```
N E X T F L O W  ~  version 0.29.0
Launching `./gsnap3.nf` [shrivelled_goldwasser] - revision: 746baced39
[warm up] executor > local
[4e/ddee1e] Submitted process > foo (1)
[52/b8c934] Submitted process > bar (1)

Received this: [@SRR636272.19519409.JA/1
GGCCCGGCAGCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTGGGGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAAGAAGGTGCCACAGATCG
+
CCCFFFFDHHD;FF=GGDHGGHIIIGHIIIBDGBFCAHG@E=6?CBDBB;?BB@BD8BB;BDB<>>;@?BB<9>&5<?288AAABDBBBBACBCAC?@AD?CAC?
, @SRR636272.19519409.JA/2
GGCCCGGCAGCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTGGGGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAAGAAGGTGCCACAGATCG
+
CCCFFFFDHHD;FF=GGDHGGHIIIGHIIIBDGBFCAHG@E=6?CBDBB;?BB@BD8BB;BDB<>>;@?BB<9>&5<?288AAABDBBBBACBCAC?@AD?CAC?
]

Received this: [@SRR636272.13995011.JB/1
GCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTGGGGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAGATCGGAAGAGCACACGTCTGAACTCC
+
BBCFDFDEFFHHFIJIHGHGHGIIFIJJJJIGGBFHHIEGBEFEFFCDDDD:@@<BB8BBDDDDDDBBB?AA?CDABDD5?CDDDBB<A<>ACBB8ACDCD@CD>
, @SRR636272.13995011.JB/2
GCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTGGGGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAGATCGGAAGAGCACACGTCTGAACTCC
+
BBCFDFDEFFHHFIJIHGHGHGIIFIJJJJIGGBFHHIEGBEFEFFCDDDD:@@<BB8BBDDDDDDBBB?AA?CDABDD5?CDDDBB<A<>ACBB8ACDCD@CD>
]

Received this: [@SRR636272.21107783.JC/1
CGGGGAGCGCGGGCCCGGCAGCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTAGGGAGCACCCCGCCGCAGGGGGACAGGCGAGATCGGAAGAGCACACGTCT
+
BCCFFDFFHHHHHJJJJJIJHHHHFFFFEEEEEEEDDDDDDBDBDBBDBBDBBB(:ABCDDDDDDDDDDDDDDDD@BBBDDDDDDDDDDDDBDDDDDDDDDDADC
, @SRR636272.21107783.JC/2
CGGGGAGCGCGGGCCCGGCAGCAGGATGATGCTCTCCCGGGCCAAGCCGGCTGTAGGGAGCACCCCGCCGCAGGGGGACAGGCGAGATCGGAAGAGCACACGTCT
+
BCCFFDFFHHHHHJJJJJIJHHHHFFFFEEEEEEEDDDDDDBDBDBBDBBDBBB(:ABCDDDDDDDDDDDDDDDD@BBBDDDDDDDDDDDDBDDDDDDDDDDADC
]

Received this: [@SRR636272.23331539.JD/1
GGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAAGAAGGTGCCACAGCTGGAGGAGCTGCTGGCCGGGAGGGACTTCACCGGCGAGATCGGAAGAG
+
CCCFFFFFHHHHHJJJJJJJJJJJJJJJHFDDBDDBDDDDDDDDDDDDADDDDDDDDDDDDDDDDDDDDDDDDDDBDBDDD9@DDDDDDDDDDDDBBDDDBDD@@
, @SRR636272.23331539.JD/2
GGAGCACCCCGCCGCAGGGGGACAGGCGGAGGAAGAAAGGGAAGAAGGTGCCACAGCTGGAGGAGCTGCTGGCCGGGAGGGACTTCACCGGCGAGATCGGAAGAG
+
CCCFFFFFHHHHHJJJJJJJJJJJJJJJHFDDBDDBDDDDDDDDDDDDADDDDDDDDDDDDDDDDDDDDDDDDDDBDBDDD9@DDDDDDDDDDDDBBDDDBDD@@
]
```
It took me quite a bit of time to arrive at that "solution".  However, now that I think about it, I'm worried that it's not really a solution at all?  From your documentation:

_The channel guarantees that items are delivered in the same order as they have been sent - but - since the process is executed in a parallel manner, there is no guarantee that they are processed in the same order as they are received. In fact, in the above example, value 3 is processed before the others._

In other words, since I'm using Nextflow processes in my "solution", it seems like maybe Nextflow can't guarantee that my FASTQ chunks will arrive at the same time (e.g. maybe chunk 2 from R1 arrives at the same time as chunk 4 from R2, and those two records would then be incorrectly stitched together)?

I obviously would much prefer that the syntax I tried at first "just worked" (i.e. the first code I referenced, where fromFilePairs() is followed by splitFastq()).  Would that be possible?  If not, is there some way you could suggest that I could continue using fromFilePairs() with splitFastq with PE reads?

My apologies if I'm missing something obvious, I'm still really new to Nextflow.  I would really appreciate any suggestions you might have.

Thank you!!
Jason

**Comment by jasonsydes** [comment]
Hrm, maybe I should add ```maxForks 1``` to my ```process bar```?  

Or maybe, I don't need to worry about it at all?  My ```process bar``` is emitting R1 and R2 at the same time (that's the first place each read pair is being split up); the next place that read pair is being processed is via the channel at splitFastq(); after that, the next place each pair is processed is at the merge() step.  Since those are all channels, and channels guarantee delivery order (unlike processes), I maybe don't need to worry?

Thank you for your assistance!!

**Comment by pditommaso** [comment]
You need to specify `flat:true` as reported in the note (blue box) in the [documentation](https://www.nextflow.io/docs/latest/operator.html#splitfastq): 

> The fromFilePairs requires the flat:true option to have the file pairs as separate elements in the produced tuples.

**Comment by pditommaso** [comment]
> In other words, since I'm using Nextflow processes in my "solution", it seems like maybe Nextflow can't guarantee that my FASTQ chunks will arrive at the same time (e.g. maybe chunk 2 from R1 arrives at the same time as chunk 4 from R2, and those two records would then be incorrectly stitched together)?

No, it means that the (parallel) execution order is not predictable but the channel sequence is consistent, therefore your example should be fine.

**Comment by jasonsydes** [comment]
Aha, this works!  Thank you!  I'm not sure how I missed that flag earlier...  Thank you again!
